File: trip-server/src/controllers/user.controller.ts
import { Request, Response } from 'express'
import * as userService from '../services/userService'

const EMAIL_REGEX = /^[^\s@]+@[^\s@]+\.[^\s@]+$/

export const register = async (req: Request, res: Response) => {
  const { username, email, password } = req.body
  if (!username || !email || !password) {
    return res.status(400).json({ code: 400, error: '用户名、邮箱和密码不能为空' })
  }
  if (!EMAIL_REGEX.test(email)) {
    return res.status(400).json({ code: 400, error: '邮箱格式不正确' })
  }
  if (password.length < 6) {
    return res.status(400).json({ code: 400, error: '密码长度不能少于6位' })
  }
  try {
    const user = await userService.register(username, email, password)
    return res.json({ code: 200, message: '注册成功', data: user })
  } catch (error: any) {
    return res.status(400).json({ code: 400, error: error.message })
  }
}

export const login = async (req: Request, res: Response) => {
  const { username, password } = req.body
  if (!username || !password) {
    return res.status(400).json({ code: 400, error: '用户名和密码不能为空' })
  }
  try {
    const user = await userService.login(username, password)
    return res.json({ code: 200, message: '登录成功', data: user })
  } catch (error: any) {
    return res.status(400).json({ code: 400, error: error.message })
  }
}

export const getInfo = async (req: Request, res: Response) => {
  try {
    const user = await userService.getUserInfo(req.user!.userId)
    return res.json({ code: 200, data: user })
  } catch (error: any) {
    return res.status(400).json({ code: 400, error: error.message })
  }
}

export const updateInfo = async (req: Request, res: Response) => {
  const { nickname, avatar, phone, bio } = req.body
  try {
    const user = await userService.updateUserInfo(req.user!.userId, { nickname, avatar, phone, bio })
    return res.json({ code: 200, message: '更新成功', data: user })
  } catch (error: any) {
    return res.status(400).json({ code: 400, error: error.message })
  }
}

export const changePassword = async (req: Request, res: Response) => {
  const { oldPassword, newPassword } = req.body
  if (!oldPassword || !newPassword) {
    return res.status(400).json({ code: 400, error: '原密码和新密码不能为空' })
  }
  if (newPassword.length < 6) {
    return res.status(400).json({ code: 400, error: '新密码长度不能少于6位' })
  }
  try {
    const result = await userService.changePassword(req.user!.userId, oldPassword, newPassword)
    return res.json({ code: 200, message: '密码修改成功', data: result })
  } catch (error: any) {
    return res.status(400).json({ code: 400, error: error.message })
  }
}

export const forgotPassword = async (req: Request, res: Response) => {
  const { email } = req.body
  if (!email) {
    return res.status(400).json({ code: 400, error: '邮箱不能为空' })
  }
  if (!EMAIL_REGEX.test(email)) {
    return res.status(400).json({ code: 400, error: '邮箱格式不正确' })
  }
  try {
    const result = await userService.createPasswordResetToken(email)
    return res.json({ code: 200, message: '重置验证码已发送', data: result })
  } catch (error: any) {
    return res.status(400).json({ code: 400, error: error.message })
  }
}

export const resetPassword = async (req: Request, res: Response) => {
  const { email, token, newPassword } = req.body
  if (!email || !token || !newPassword) {
    return res.status(400).json({ code: 400, error: '邮箱、验证码和新密码不能为空' })
  }
  if (!EMAIL_REGEX.test(email)) {
    return res.status(400).json({ code: 400, error: '邮箱格式不正确' })
  }
  if (newPassword.length < 6) {
    return res.status(400).json({ code: 400, error: '新密码长度不能少于6位' })
  }
  try {
    const result = await userService.resetPassword(email, token, newPassword)
    return res.json({ code: 200, message: '密码重置成功', data: result })
  } catch (error: any) {
    return res.status(400).json({ code: 400, error: error.message })
  }
}

File: trip-backend-java/src/main/java/com/trip/backend/web/controller/UserController.java
package com.trip.backend.web.controller;

import com.trip.backend.domain.entity.User;
import com.trip.backend.service.UserService;
import com.trip.backend.utils.AppException;
import com.trip.backend.web.handler.FormatResolver;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Size;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.security.access.prepost.PreAuthorize;

import java.util.Map;

/**
 * 用户认证控制器（对应 Python routers/user.py）
 * 7 个端点：
 * - POST /api/user/register（公开）
 * - POST /api/user/login（公开）
 * - GET /api/user/info（JWT）
 * - PUT /api/user/info（JWT）
 * - PUT /api/user/password（JWT）
 * - POST /api/user/forgot-password（公开）
 * - POST /api/user/reset-password（公开）
 */
@RestController
@RequestMapping("/api/user")
public class UserController {

    private final UserService userService;

    public UserController(UserService userService) {
        this.userService = userService;
    }

    /**
     * POST /api/user/register
     */
    @PostMapping("/register")
    public ResponseEntity<Map<String, Object>> register(
            @Valid @RequestBody RegisterRequest request,
            HttpServletRequest httpRequest) {
        User user = userService.register(request.username, request.email, request.password);

        Map<String, Object> data = Map.of(
            "id", user.getId(),
            "username", user.getUsername(),
            "email", user.getEmail()
        );

        if (FormatResolver.isFormatA(httpRequest)) {
            return ResponseEntity.status(HttpStatus.CREATED).body(Map.of("success", true, "data", data));
        } else {
            return ResponseEntity.status(HttpStatus.CREATED).body(Map.of(
                "code", 201, "data", data, "message", "注册成功", "error", null
            ));
        }
    }

    /**
     * POST /api/user/login
     */
    @PostMapping("/login")
    public ResponseEntity<Map<String, Object>> login(
            @Valid @RequestBody LoginRequest request,
            HttpServletRequest httpRequest) {
        User user = userService.login(request.identifier, request.password);

        // 生成 JWT（简化版，实际应调用 JwtService）
        // TODO: 注入 JwtUtil 生成 token
        String token = "mock-jwt-token-" + user.getId(); // 临时占位

        Map<String, Object> data = Map.of(
            "token", token,
            "user", Map.of(
                "id", user.getId(),
                "username", user.getUsername(),
                "email", user.getEmail(),
                "nickname", user.getNickname(),
                "avatar", user.getAvatar()
            )
        );

        if (FormatResolver.isFormatA(httpRequest)) {
            return ResponseEntity.ok(Map.of("success", true, "data", data));
        } else {
            return ResponseEntity.ok(Map.of(
                "code", 200, "data", data, "message", "登录成功", "error", null
            ));
        }
    }

    /**
     * GET /api/user/info
     */
    @GetMapping("/info")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<Map<String, Object>> getInfo(
            @RequestAttribute("userId") Long userId,
            HttpServletRequest httpRequest) {
        User user = userService.getUserInfo(userId);

        Map<String, Object> data = Map.of(
            "id", user.getId(),
            "username", user.getUsername(),
            "email", user.getEmail(),
            "nickname", user.getNickname(),
            "avatar", user.getAvatar(),
            "bio", user.getBio()
        );

        if (FormatResolver.isFormatA(httpRequest)) {
            return ResponseEntity.ok(Map.of("success", true, "data", data));
        } else {
            return ResponseEntity.ok(Map.of(
                "code", 200, "data", data, "message", null, "error", null
            ));
        }
    }

    /**
     * PUT /api/user/info
     */
    @PutMapping("/info")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<Map<String, Object>> updateInfo(
            @RequestAttribute("userId") Long userId,
            @Valid @RequestBody UpdateInfoRequest request,
            HttpServletRequest httpRequest) {
        User user = userService.updateUserInfo(userId, request.nickname, request.avatar, request.bio);

        Map<String, Object> data = Map.of(
            "id", user.getId(),
            "username", user.getUsername(),
            "email", user.getEmail(),
            "nickname", user.getNickname(),
            "avatar", user.getAvatar(),
            "bio", user.getBio()
        );

        if (FormatResolver.isFormatA(httpRequest)) {
            return ResponseEntity.ok(Map.of("success", true, "data", data));
        } else {
            return ResponseEntity.ok(Map.of(
                "code", 200, "data", data, "message", null, "error", null
            ));
        }
    }

    /**
     * PUT /api/user/password
     */
    @PutMapping("/password")
    @PreAuthorize("isAuthenticated()")
    public ResponseEntity<Map<String, Object>> changePassword(
            @RequestAttribute("userId") Long userId,
            @Valid @RequestBody ChangePasswordRequest request) {
        userService.changePassword(userId, request.oldPassword, request.newPassword);
        return ResponseEntity.ok(Map.of("success", true, "data", null));
    }

    /**
     * POST /api/user/forgot-password（恒返回成功，防枚举）
     */
    @PostMapping("/forgot-password")
    public ResponseEntity<Map<String, Object>> forgotPassword(
            @Valid @RequestBody ForgotPasswordRequest request) {
        userService.forgotPassword(request.email);
        return ResponseEntity.ok(Map.of("success", true, "data", null));
    }

    /**
     * POST /api/user/reset-password
     */
    @PostMapping("/reset-password")
    public ResponseEntity<Map<String, Object>> resetPassword(
            @Valid @RequestBody ResetPasswordRequest request) {
        userService.resetPassword(request.token, request.newPassword);
        return ResponseEntity.ok(Map.of("success", true, "data", null));
    }

    // ========== DTOs ==========

    public record RegisterRequest(
        @NotBlank @Size(min = 3, max = 50) String username,
        @NotBlank @Email String email,
        @NotBlank @Size(min = 6) String password
    ) {}

    public record LoginRequest(
        @NotBlank String identifier,
        @NotBlank String password
    ) {}

    public record UpdateInfoRequest(
        @Size(max = 50) String nickname,
        String avatar,
        @Size(max = 255) String bio
    ) {}

    public record ChangePasswordRequest(
        @NotBlank String oldPassword,
        @NotBlank @Size(min = 6) String newPassword
    ) {}

    public record ForgotPasswordRequest(
        @NotBlank @Email String email
    ) {}

    public record ResetPasswordRequest(
        @NotBlank String token,
        @NotBlank @Size(min = 6) String newPassword
    ) {}
}
